audio-worklet/noisegate-test.html

```diff
@@ -0,0 +1,451 @@
+          let noiseGateOutput = buffer.getChannelData(0);
+          let originalOutput = buffer.getChannelData(1);
+         
+          const closedIndex =
+              toIndex_(timeToThreshold, sampleRate) + scriptProcessorLatency;
+
+          // The output is divided into a latency phase, an unaltered phase, and
+          // a closed phase.
+          let latencyPhase = noiseGateOutput.slice(0, scriptProcessorLatency);
+          let unalteredPhase =
+              noiseGateOutput.slice(scriptProcessorLatency, closedIndex);
+          let unalteredPhaseReference =
+              originalOutput.slice(scriptProcessorLatency, closedIndex);
+
+          // The first samples run through the noise gate will be muted because
+          // of envelope delay.
+          const openingDelay =
+              getFirstIndexBeforeRemainingSamplesMatch_(
+                  unalteredPhase, unalteredPhaseReference);
+          
+          // The observed delay of the envelope is 0.002685546875s, and changes
+          // depending on choice of time constant. Unlike the test above, the
+          // delay cannot be easily predicted since the envelope measures
+          // a sine wave.
+          const delayInSeconds = toTime_(openingDelay, sampleRate);
+          should(delayInSeconds,
+                 'The envelope delay for the very first samples')
+              .beCloseTo(0.0026, {threshold: 0.1});
+
+          unalteredPhase = unalteredPhase.slice(openingDelay,
+                                                unalteredPhase.length);
+          unalteredPhaseReference = unalteredPhaseReference.slice(
+              openingDelay, unalteredPhaseReference.length);
+
+          // Based on knowledge of when the signal goes beneath the threshold,
+          // it is possible to predict when the signal should be muted. However,
+          // this calculation is imperfect since the envelope detection
+          // algorithm has a delay.
+          let ideallyClosedPhase =
+              noiseGateOutput.slice(closedIndex, buffer.length);
+          
+          // The difference between when the gate is expected to be closed and
+          // is closed) is measured at 0.002197265625s. If settings change, then
+          // closeDelay can be inspected and the assertions can compare to this
+          // benchmark instead. The delay occurs due to envelope lag and
+          // changes based on timeConstant. Slight delay may be problematic if
+          // the noise gate is intended to immediately mute sound
+          // upon hitting a threshold and less problematic if a smooth
+          // transition to silence is desired.
+          let closeDelay =
+              getFirstIndexBeforeRemainderIs_(ideallyClosedPhase, 0);
+          should(
+              toTime_(closeDelay, sampleRate),
+              'The delay between when the envelope is expected to be closed and'
+              + ' when it is closed')
+              .beCloseTo(0.0022, {threshold: 0.1});
```

Comment:
Here the relative error at 2^18 is 0.0138. Would it be better to set each threshold to be at the conservative 5%? Or to set each threshold to be as small as it can be.